src/ssl.c

```diff
@@ -3448,6 +3448,207 @@ void wolfSSL_CertManagerFree(WOLFSSL_CERT_MANAGER* cm)
 
 }
 
+#if defined(OPENSSL_EXTRA) && !defined(NO_FILESYSTEM)
+#if defined(WOLFSSL_SIGNER_DER_CERT)
+/******************************************************************************
+* wolfSSL_CertManagerGetCerts - retrieve stack of X509 certificates in a
+* certificate manager (CM).
+*
+* RETURNS:
+* returns stack of X509 certs on success, otherwise returns a NULL.
+*/
+WOLFSSL_STACK* wolfSSL_CertManagerGetCerts(WOLFSSL_CERT_MANAGER* cm)
+{
+    WOLFSSL_STACK* sk = NULL;
+    Signer* signers = NULL;
+    word32  row = 0;
+    DecodedCert* dCert = NULL;
+    WOLFSSL_X509* x509 = NULL;
+    int found = 0;
+
+    if (cm == NULL)
+        return NULL;
+
+    sk = wolfSSL_sk_X509_new();
+
+    if (sk == NULL) {
+        return NULL;
+    }
+
+    if (wc_LockMutex(&cm->caLock) != 0) {
+        goto error_init;
+    }
+
+    for (row = 0; row < CA_TABLE_SIZE; row++) {
+        signers = cm->caTable[row];
+        while (signers && signers->derCert && signers->derCert->buffer) {
+
+            dCert = (DecodedCert*)XMALLOC(sizeof(DecodedCert), cm->heap,
+                                          DYNAMIC_TYPE_DCERT);
+            if (dCert == NULL) {
+                goto error;
+            }
+
+            XMEMSET(dCert, 0, sizeof(DecodedCert));
+
+            InitDecodedCert(dCert, signers->derCert->buffer,
+                            signers->derCert->length, cm->heap);
+
+            /* Parse Certificate */
+            if (ParseCert(dCert, CERT_TYPE, NO_VERIFY, cm)) {
+                goto error;
+            }
+
+            x509 = (WOLFSSL_X509*)XMALLOC(sizeof(WOLFSSL_X509), cm->heap,
+                    DYNAMIC_TYPE_X509);
+
+            if (x509 == NULL) {
+                goto error;
+            }
+
+            InitX509(x509, 1, NULL);
+
+            if (CopyDecodedToX509(x509, dCert) == 0) {
+
+                if (wolfSSL_sk_X509_push(sk, x509) != SSL_SUCCESS) {
+                    WOLFSSL_MSG("Unable to load x509 into stack");
+                    goto error;
+                }
+            }
+            else {
+                goto error;
+            }
+
+            found = 1;
+
+            signers = signers->next;
+
+            FreeDecodedCert(dCert);
+            XFREE(dCert, cm->heap, DYNAMIC_TYPE_DCERT);
+            dCert = NULL;
```

Comment:
Why isn't x509 marked NULL as well?  Think about what happens after 1 is found but a later one fails on dCert malloc for example.

Reply:
@toddouska In that error case, x509 will be freed by wolfSSL_sk_X509_free() at the end. Same with your other comments. Let me know if I missed something?

Reply:
x509 will also be freed by FreeX509 (line 3549) in that case which is a double free.

Reply:
The list of allocated X509's is what is returned, which is why its not free'd here. It only gets free'd in the error case. The caller must free using `wolfSSL_sk_X509_free(sk);`. Looks to me like the cleanup cases are handled correctly. I do think the "goto" cases and cleanup logic should be improved... its confusing.

Reply:
@toddouska,  Good point but It's not a double-free. FreeX509 (line 3549)  is there for readability and consistency. wolfSSL_sk_X509_free() correctly handles if x509 has been freed. I can remove it if needed.

Reply:
I'm not sure that's true.  How can you guarantee that calling FreeX509() on the x509 pointer and wolfSSL_sk_X509_free() with the same pointer (which will call FreeX509()) won't cause a problem now or in the future when you can easily avoid the issue?

Reply:
@toddouska, Removed. Actually you are correct and it should be removed.